particle-iot/device-os
# Monitor mode debug on Gen3 device

### Problem

Implement monitor mode debug to support debugging in a BLE connection.

#### Note
**This PR is under development.**

TODO:
1. Should we optimize makefile to support compile assembly file
1. Optimize BLE control request to prevent the disconnection operation while stopping at a break point.
1. Only tested on MacOS

### Steps to Test

1. Comment the content of `ble_disconnect()` in `hal/src/nRF52840/ble_hal.cpp` in case BLE driver initiates to terminate the BLE connection. This is a workaround for **TODO_#2**.
1. Build firmware `make clean all PLATFORM=xenon TEST=app/blank DEBUG_BUILD=y MODULAR=n`
1. Flash the blank application to a Xenon
1. Start debugging on VSCode
1. Send GDB command of monitor mode debug  via DEBUG CONSOLE
```
mon reset 0
mon exec SetMonModeDebug=1
mon exec SetMonModeVTableAddr=0x30000
```
![image](https://user-images.githubusercontent.com/7424522/54477776-edbc6180-4845-11e9-8e57-41f644e6f75e.png)
1. Press `Continue (F5)` to run
1. Use Lightblue to connect to the Xenon
1. Press `Pause (F6)` to stop
1. Observe whether the connection is terminated

### Example App

### References
[CH29931]

[Monitor Mode Debugging in Segger embedded studios](https://github.com/NordicPlayground/j-link-monitoring-mode-debugging/blob/master/pca10040/s132/ses/README.md)

---

### Completeness
